Proposal: `runtime.onInvalidated` event (#792)
hober/web-extensions: proposals/runtime_on_invalidated.md

# Proposal: browser.runtime.onInvalidated event

**Summary**

`browser.runtime.onInvalidated`

> This event allows extension contexts which stay alive to be notified when the
> extension context get invalidated.

Reviewer: ```suggestion
This event notifies extension contexts that remain alive when the extension context gets invalidated.
```

**Document Metadata**

**Author:** carlosjeurissen

**Sponsoring Browser:** Chrome

**Contributors:** bershanskiy

**Created:** 2025-03-26

**Related Issues:** https://github.com/w3c/webextensions/issues/138

## Motivation

### Objective

> This event allows extension contexts which stay alive to be notified when the

Reviewer 2: "extension context" is very generic, while this seems to primarily be targeting content scripts. Can we restrict this to content scripts?

Author: @Rob--W it is generic on purpose as to cover both content scripts and iframed extension contexts on third party websites. These iframes stay alive on some browsers and require this event to handle their state. The objective states `contexts which stay alive` to exclude extension contexts which get killed on invalidation.

Updated the objective to be more explicit on this behaviour.

> This event allows extension contexts which stay alive to be notified when the
> extension context get invalidated.

Reviewer: ```suggestion
This event notifies extension contexts that remain alive when the extension context gets invalidated.
```

#### Use Cases

When the extension context gets invalidated, this means extension contexts

> which stay alive can no longer use the chrome APIs or communicate with other

Reviewer: ```suggestion
which stay alive can no longer use the extension APIs or communicate with other
```

extension contexts.

For extension contexts, knowing when this happens is crucial for moving
forward. Alternatives can be used in this situation, like `window.open` instead
of `browser.tabs.create`. If moving forward is no option, this event can be
used for initiating the cleanup of any side effects caused in the frames
extensions have been operating in.

### Known Consumers

Any extension with a content script which applies certain side effects to the
web page.

## Specification

### Schema

New onInvalidated event:
```json
{
  "name": "onInvalidated",
  "type": "function",

>   "description": "Fired when the extension context get invalidated.",

Reviewer 2: ```suggestion
  "description": "Fired when the extension context get invalidated while the underlying document remains valid.",
```

Suggested tweak to the description. Otherwise you would also have to fire the event on page unload and browser shutdown.

This description also precludes non-content scripts. If you want to include non-content scripts I'd like to see a good justification. Regular document have the pagehide, beforeunload and unload events already. Background contexts have runtime.onSuspend.

  "parameters": [
    {
      "type": "object",
      "name": "details",
      "properties": {
        "reason": {
          "$ref": "OnInvalidatedReason",
          "description": "The reason that this event is being dispatched."
        }
      }
    }
  ]
}
```

The event should have a reason of type OnInvalidatedReason

OnInvalidatedReason:

```json
{
  "id": "OnInvalidatedReason",
  "type": "string",
  "enum": [
    {"name": "uninstall", "description": "Specifies the event reason as an uninstallation."},
    {"name": "update", "description": "Specifies the event reason as an extension update."},
    {"name": "reload", "description": "Specifies the event reason as an extension reloading."},
    {"name": "disable", "description": "Specifies the event reason as an extension disabling."}
  ],
  "description": "The reason that this event is being dispatched."
}
```

### Behavior

The event will be fired when the extension context get invalidated.
The event comes with a reason which specifies why an extension got invalidated.

### New Permissions

No new permissions introduced.

### Manifest File Changes

No new manifest fields.

## Security and Privacy

### Exposed Sensitive Data

> Document any sensitive data or personally-identifiable information the API

Reviewer 2: This is unmodified from the template. Replace it please.

Author: Fixed.

exposes to the extension.

### Abuse Mitigations

Currently extensions could try checking if an API gets thrown with an
invalidated error. This would allow an extension to figure out if the extension
is invalidated. No new abuse surface is introduced.

As for the returned reason. The `uninstall` reason can not be abused more than

> currently `browser.runtime.setUninstalledURL`.

Reviewer 2: There is a distinction between the uninstall URL and the event: the URL is opened in a new tab, whereas the event is dispatched while the extension still has access to other pages. In theory it could modify every page to nag the user to undo the uninstallation. Is this too contrived or should we call that out?

Author: @Rob--W Good catch. Updated the PR to mention this scenario.

# Proposal: browser.runtime.onInvalidated event>

> ** How to Use This Template **

Reviewer 3: Can we drop this?

Author: Dropped. Thanks.

Reviewer 4: Doesn't look dropped yet in the latest PR

Author: Thanks. It is dropped now for real)

See [Proposal Process](proposal_process.md) for the detailed process on how to
propose new APIs and use this template.  Each section includes instructions on
what to include.  Delete the instructions for a given section once it's filled
out.  Remove this section once the template is filled out.

**Summary**

`browser.runtime.onInvalidated`

> This event allows extension frames and contentScripts to be notified when

Reviewer 3: ```suggestion
This event allows extension frames and content scripts to be notified when
```

...and the same throughout? I think we always use this rather than lower camelcase.

Author: Updated. Thanks.

background scripts get invalidated.

**Document Metadata**

**Author:** carlosjeurissen

**Sponsoring Browser:** Chrome

**Contributors:** TBD

**Created:** 2025-03-26

**Related Issues:** https://github.com/w3c/webextensions/issues/138

## Motivation

### Objective

This event allows extension frames and contentScripts to be notified when
background scripts get invalidated.

#### Use Cases

> Cleanup of effects in contentScripts. After-invalidated handling of frames.

Reviewer 3: It would be nice to elaborate on these. I know what you mean, but it would unclear to someone else why `browser.tabs` wouldn't be available after the background went away, for example :)

Author: Sure. Elaborated a bit more in the proposal.

Reviewer 3: Hmm, it looks the same still?

Author: @oliverdunk try a refresh

Reviewer 3: Much better! Thanks.

For example, know when to use browser.tabs vs window.open.

### Known Consumers

Any extension with a content script which applies certain side effects to the
web page.

## Specification

### Schema

New onInvalidated event:
```json
{
  "name": "onInvalidated",
  "type": "function",

>   "description": "Fired when the extension background scripts get invalidated.",

Reviewer 5: Chrome has always called it "context invalidated" so I think it makes sense to keep calling it that, rather than referring to background scripts/workers. Extensions without a background worker are also affected (e.g. content script attempts to message the sidePanel or some `chrome-extensions://**/index.html` page)

Author: @fregante changed to context invalidated. Thanks!

  "parameters": [
    {
      "type": "object",
      "name": "details",
      "properties": {
        "reason": {

>           "optional": true,

Reviewer 3: Why is this optional?

Author: @oliverdunk browsers may not want to expose this for either partial implementation or for security / privacy reasons. See abuse / security parts.

Reviewer 3: Did anyone express that they would do this? I'd rather have a consistent set of properties unless there's a compelling reason otherwise.

Author: @oliverdunk I do not believe any browser vendor stated any kind of in favour / opposed for the reason property. So this is left purposely open. Talking of which, would Chrome be in favour / opposed for the reason property?

Reviewer 3: We're supportive. As discussed, we'd prefer to remove the `optional` modifier here.

Author: @oliverdunk updated to not be optional.

          "$ref": "OnInvalidatedReason",
          "description": "The reason that this event is being dispatched."
        }
      }
    }
  ]
}
```

Optionally gives a reason of type OnInvalidatedReason

OnInvalidatedReason:

```json
{
  "id": "OnInvalidatedReason",
  "type": "string",
  "enum": [
    {"name": "uninstalled", "description": "Specifies the event reason as an uninstallation."},

>     {"name": "update", "description": "Specifies the event reason as an extension update."},

Reviewer 5: While it's a dev-only event, does "reload" fall into this one?

Author: Added reload as reason. Love to hear from browsers on their thoughts on this.

Reviewer 4: I like it.  It's also not (necessarily) dev-only.  There are other times extensions could be reloaded, and extensions can trigger the reload themselves via runtime.reload().

# Proposal: browser.runtime.onInvalidated event>

** How to Use This Template **

See [Proposal Process](proposal_process.md) for the detailed process on how to
propose new APIs and use this template.  Each section includes instructions on
what to include.  Delete the instructions for a given section once it's filled
out.  Remove this section once the template is filled out.

**Summary**

`browser.runtime.onInvalidated`

> This event allows extension frames and content scripts to be notified when

Reviewer 4: extension frames should be either closed or terminated when the extension is unloaded, at least in Chrome.  Is that not the case in other browsers?  Or should we indicate this only applies for content scripts?

Author: Not all frames may be closed by all browsers. I can change this to extension context to be more inclusive? What do you think @rdcronin ?

Author: @rdcronin updated to use extension contexts.

### Objective

This event allows extension frames and content scripts to be notified when
context get invalidated.

#### Use Cases

When the extension context gets invalidated, this means the content script
or frame can no longer use the chrome APIs or communicate with the extension.

As content script or frame, knowing when this happens is crucial for moving
forward. Alternatives can be used in this situation, like `window.open` instead
of `browser.tabs.create`. If moving forward is no option, this event can be
used for initiating the cleanup of any side effects caused in the frames
extensions have been operating in.

### Known Consumers

Any extension with a content script which applies certain side effects to the
web page.

## Specification

### Schema

New onInvalidated event:
```json
{
  "name": "onInvalidated",
  "type": "function",
  "description": "Fired when the extension context get invalidated.",
  "parameters": [
    {
      "type": "object",
      "name": "details",
      "properties": {
        "reason": {
          "$ref": "OnInvalidatedReason",
          "description": "The reason that this event is being dispatched."
        }
      }
    }
  ]
}
```

> Optionally gives a reason of type OnInvalidatedReason

Reviewer 4: as discussed yesterday, this shouldn't be optional

Author: Fixed.

OnInvalidatedReason:

```json
{
  "id": "OnInvalidatedReason",
  "type": "string",
  "enum": [

>     {"name": "uninstalled", "description": "Specifies the event reason as an uninstallation."},

Reviewer 4: nit: we're mixing tenses -- "uninstalled" and "update" (I would expect "uninstall" and "update" or "uninstalled" and "updated")

Author: Thanks. Resolved.

    {"name": "update", "description": "Specifies the event reason as an extension update."},
    {"name": "reload", "description": "Specifies the event reason as an extension reloading."},
    {"name": "disabled", "description": "Specifies the event reason as an extension disabling."}
  ],
  "description": "The reason that this event is being dispatched."
}
```

### Behavior

> The event would be fired when the extension context get invalidated.

Reviewer 4: ```suggestion
The event will be fired when the extension context get invalidated.
```

(or "is")

Author: Thanks!
